### PR TITLE
[FIX #2182] Add Access-Control-Allow-Credentials for CORS settings 

### LINKED ISSUE #2182: Add Access-Control-Allow-Credentials for CORS settings
# This is a Feature Proposal
## Description

Right now I am developing a cross origin resource sharing (CORS) application which requires sharing cookies. The server is required to return the Access-Control-Allow-Credentials header set to true to allow this. While there is way to set headers, methods and origins with serverless, there is no easy way of setting the allow-credentials header. Will have to go to aws console to enable this which is a headache for maintenance. 
## Additional Data
- **_Serverless Framework Version you're using**_:
- **_Operating System**_:
- **_Stack Trace**_:
- **_Provider Error messages**_:

### PR DESCRIPTION
**Pull request by fruffin**
## What did you implement:

**_Implementing Issue:**_ #2182 

This PR adds support for a new "allowCredentials" flag in the CORS configuration for API Gateway events
## How did you implement it:

The implementation is fairly trivial. I just ended up extending the CORS configuration object to add the new flag and added the extra `Access-Control-Allow-Credentials` header to preflight requests when CORS is enabled.

If the flag is not specified in configuration, it will default to "true"
### Scenario 1
- Start a project with a function responding to an API Gateway event
- Enable CORS with the old configuration
- Deploy
- Verify that the preflight request has the `Access-Control-Allow-Credentials` header set to `true`
### Scenario 2
- Start a project with a function responding to an API Gateway event
- Enable CORS with the old configuration
- Add the `allowCredentials` configuration element to CORS section of the API Gateway event and set it to `true`
- Deploy
- Verify that the preflight request has the `Access-Control-Allow-Credentials` header set to `true`
### Scenario 3
- Start a project with a function responding to an API Gateway event
- Enable CORS with the old configuration
- Add the `allowCredentials` configuration element to CORS section of the API Gateway event and set it to `false`
- Deploy
- Verify that the preflight request has the `Access-Control-Allow-Credentials` header set to `false`

Alternatively, there are some unit tests verifying these scenarios
## Todos:
- [X] Write tests
- [X] Write documentation
- [X] Fix linting errors
- [X] Make sure code coverage hasn't dropped
- [X] Provide verification config/commands/resources
- [x] Change ready for review message below

**_Is this ready for review?:**_ YES

### CODE REVIEW

**Comment by fruffin** [comment]
First build failed due to linting issue in CLI.js. 

All good now.

**Comment by flomotlik** [comment]
@fruffin can you rebase? We fixed the issue in the CLI file in a separate branch, sorry for that.

**Comment by fruffin** [comment]
@flomotlik yeah I just saw that. I got rid of an additional `'` which is why there is a conflict. I'll rebase now.
Will post here when done

**Comment by fruffin** [comment]
@flomotlik done. I reverted my last commit and did a fresh merge of the upstream. Looks like all is good in the world again :)
Good to go when you're ready

**Comment by str3tch** [comment]
I'm not sure if this is complete to get credentials working completely. With this PR I get the credentials flag set correctly in the OPTIONS but the GET request does not have it set. 

I will investigate further...

**Comment by fruffin** [comment]
@str3tch My understanding was that it is only necessary for pre-flight requests

**Comment by str3tch** [comment]
Hi @fruffin, it's needed on all requests, not just OPTIONS (or at least that's what my browser is telling me.

See below for my subsequent failed GET request after a successful OPTIONS:

image

**Comment by str3tch** [comment]
I think this change might fix it:

methods.js, line 537:
```
         // 4. Has CORS enabled?
          if (hasCors(event)) {
            corsPreflight[path] = generateCorsPreflightConfig(event.http.cors,
              corsPreflight[path], method);

            let allowCredentials = true;
            if (event.http.cors.allowCredentials !== 'undefined') {
              allowCredentials = event.http.cors.allowCredentials;
            }  

            const corsHeader = {
              'Access-Control-Allow-Origin':
              `'${corsPreflight[path].origins.join('\',\'')}'`,
              'Access-Control-Allow-Credentials': `'${allowCredentials}'`,
            };

            _.merge(methodResponseHeaders, corsHeader);
            _.merge(integrationResponseHeaders, corsHeader);
          }
```

**Comment by fruffin** [comment]
That's odd, it works for me in the browser and I'm just returning the header for OPTIONS requests.
I had the exact same error message in my react app before, which is why I ended up working on this. I do still have the code to have it on all requests, but I took it out because it didn't seem necessary.

It turns out you're right though: https://developer.mozilla.org/en-US/docs/Web/HTTP/Access_control_CORS#Access-Control-Allow-Credentials

I'll bring back my change tonight.
@flomotlik is it possible to change the approval status back to "not approved" for now? It looks like another round of review will be necessary. Sorry about that...

**Comment by str3tch** [comment]
Awesome thanks Fabien. I've left the office now so can't continue testing
but I think this is the right thing to do.

**Comment by fruffin** [comment]
New code and tests committed.
It's all for the better in the end as I found a better way of making the original change in the process :)

@flomotlik Good to go again. Sorry about the mess...

**Comment by str3tch** [comment]
reaaaaaly need this to get our application running. Is it slated for the next release, @flomotlik ?

cheers :)

**Comment by fruffin** [comment]
ok, looks like I can start over, now that everything has changed again...

**Comment by asprouse** [comment]
I needed to get this working for a project. I have adapted your change to the latest code in my fork https://github.com/asprouse/serverless. I can make a PR if you guys want or you can grab the changes from my repo. Either way it needs to be reviewed

**Comment by str3tch** [comment]
I am using Fabien's current PR to get my project working.

however I note that Serverless now suggest to use lambda-proxy, and setup
CORS stuff in the code, not in the config. Have you guys tried that route?

**Comment by asprouse** [comment]
@str3tch as far as I can tell even when you are using lambda proxy you still need to add the correct headers into your response object (this is mentioned in the docs). For example:
```javascript
const response = {
  statusCode: 200,
  body: JSON.stringify(body),
  headers: {
    'Access-Control-Allow-Origin': process.env.ALLOW_ORIGIN,
    'Access-Control-Allow-Credentials': true
  }
};

callback(null, response);
```
If you are using the lamdba integration you could put these headers into your custom response template. 

@fruffin let me know if my branch is of any use.

**Comment by str3tch** [comment]
@asprouse yes you're right, but @fruffin and my original changes were to change serverless code to get the credentials headers returned. But now with `lambda-proxy` integration, looks like we might not need to change serverless code, we can just add the headers in the response object in our own code (not serverless's code)

However, if you don't use `lambda-proxy` and just use `lambda`, there will need to be serverless changes

**Comment by asprouse** [comment]
@str3tch lambda-proxy just sets the integration type in the CloudFormation template  to AWS_PROXY:
```json
"ApiGatewayMethodGraphqlPost": {
      "Type": "AWS::ApiGateway::Method",
      "Properties": {
        "HttpMethod": "POST",
        "RequestParameters": {},
        "ResourceId": {
          "Ref": "ApiGatewayResourceGraphql"
        },
        "RestApiId": {
          "Ref": "ApiGatewayRestApi"
        },
        "AuthorizationType": "NONE",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:aws:apigateway:",
                {
                  "Ref": "AWS::Region"
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "GraphqlLambdaFunction",
                    "Arn"
                  ]
                },
                "/invocations"
              ]
            ]
          }
        },
        "MethodResponses": []
      }
    },
```
The CORS configuration is independant of either integration as  OPTIONS request is handled by a Mock integration type so that you never invoke the lamdba function:
```
"ApiGatewayMethodGraphqlOptions": {
      "Type": "AWS::ApiGateway::Method",
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "OPTIONS",
        "MethodResponses": [
          {
            "StatusCode": "200",
            "ResponseParameters": {
              "method.response.header.Access-Control-Allow-Origin": true,
              "method.response.header.Access-Control-Allow-Headers": true,
              "method.response.header.Access-Control-Allow-Methods": true,
              "method.response.header.Access-Control-Allow-Credentials": true
            },
            "ResponseModels": {}
          }
        ],
        "RequestParameters": {},
        "Integration": {
          "Type": "MOCK",
          "RequestTemplates": {
            "application/json": "{statusCode:200}"
          },
          "IntegrationResponses": [
            {
              "StatusCode": "200",
              "ResponseParameters": {
                "method.response.header.Access-Control-Allow-Origin": "'https://dev.example.com'",
                "method.response.header.Access-Control-Allow-Headers": "'Content-Type,X-Amz-Date,Authorization,X-Api-Key,X-Amz-Security-Token'",
                "method.response.header.Access-Control-Allow-Methods": "'OPTIONS,POST'",
                "method.response.header.Access-Control-Allow-Credentials": "'true'"
              },
              "ResponseTemplates": {
                "application/json": ""
              }
            }
          ]
        },
        "ResourceId": {
          "Ref": "ApiGatewayResourceGraphql"
        },
        "RestApiId": {
          "Ref": "ApiGatewayRestApi"
        }
      }
    },
```
TLDR; the CORS config is the same for both integrations.

**Comment by fruffin** [comment]
After waiting a month for this PR to be approved and merged, the code has now changed significantly.

@asprouse picked up this PR and ported the changes to the new version of the API Gateway integration plugin in the new PR #2736 .

I'm not going to copy his changes here, but I'm not closing this PR yet either to keep the conversation history until @asprouse PR gets merged

**Comment by pmuens** [comment]
Hey @fruffin thank you very much for this PR! 👍  🎉 

We're very sorry that this was not reviewed in time and was outdated quite frequently 😢
This is frustrating and shouldn't happen again anytime soon!
We've polished our processes a little bit and will try to be way more responsive so that PRs with such changes get into `master` way faster.

Again. Thank you very much for the time you've put into this one! Really appreciate all your contributions you've done so far! 💯 

Thanks @asprouse for lending a helping hand here! We'll review the changes ASAP. I'll close this PR now as we should continue the discussion in #2736.

**Comment by fruffin** [comment]
@pmuens That's ok. At least thanks to @asprouse it wasn't wasted effort.
Any chance to get my other PR (#2434) looked at some time soon then :) ? It's been sitting there for over a month as well. 

Thanks

**Comment by pmuens** [comment]
@fruffin for sure! We've already looked into it and will continue with the next step (code review) soon! Thanks for the PR btw! 👍
